### PR TITLE
feat: add parent model and service for renting/returning library items

### PR DESCRIPTION
**Pull request by blackspyek**
# **Feature**: Models and services for renting/returning library items

## Summary
This PR introduces a new feature for managing library items, with support for renting and returning items. It refactors the current system by adding a parent `LibraryItem` model, implementing concrete models like `Book` and `Movie`, and introducing services for both book management and rental operations. It also includes new endpoints for interacting with these items, as well as several improvements to the codebase.

## Key Changes:
1. **LibraryItem Model**:
   - Created a parent `LibraryItem` model with shared properties and methods.
   - Implemented specific models: `Book` and `Movie`, both extending `LibraryItem` and adding item-specific fields and logic.

2. **Interfaces**:
   - Introduced interfaces like `hasAuthor` to handle common methods across models, ensuring consistency and reusability of code.

3. **Book Service**:
   - Developed the `BookService` to handle book-specific logic, such as creation, retrieval, and updates.

4. **Rental Module**:
   - Created the `RentalService` to manage renting/returning library items.
   - Added `RentalPolicyService` to handle rental policies (e.g., availability, rental period).

5. **Endpoints**:
   - **BookController**:
     - `POST /api/book` with `BookCreateDto` (Create a new book).
     - `GET /api/book/{id}` (Get book by ID).
     - `GET /api/book/author/{author}` (Get books by author).
     - `GET /api/book/genre/{genre}` (Get books by genre).
     - `GET /api/book/isbn/{isbn}` (Get book by ISBN).
     - `PUT /api/{id}` with `BookCreateDto` (Update a book).
     - `DELETE /api/{id}` (Delete a book).

   - **RentalController**:
     - `GET /api/rentals/user/{userId}` (Get all rentals for a user).
     - `GET /api/rentals/available` (Get all available items for rent).
     - `POST /api/rentals/rent` with `RentDto` (Rent an item).
     - `POST /api/rentals/return/{itemId}` (Return a rented item).

6. **Code Improvements**:
   - Fixed small bugs and addressed warnings from the linter to improve code quality.

7. **Testing**:
   - Manually tested all the newly created endpoints to ensure functionality.

## Next Steps:
- Review the new models and services for potential improvements.
- Introduce MovieController and implement MovieService

## Notes:
- This is a significant change that introduces several new modules and endpoints. Ensure that the new structure aligns with future features and operations.

### CODE REVIEW

File: library/src/main/java/org/pollub/library/rental/service/RentalService.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    private final IUserService userService;
+    private final IRentalPolicy rentalPolicy;
+
+    @Override
+    public LibraryItem rentItem(RentDto rentDto) {
+        LibraryItem libraryItem = getLibraryItemOrThrow(rentDto.getLibraryItemId());
+
+        User user = userService.findById(rentDto.getUserId());
+        validateAbilityToRentOrThrow(user.getId(), libraryItem);
+        setLibraryItemToBeRatedByUser(libraryItem, user);
```
Comment (DFraczek) [inline library/src/main/java/org/pollub/library/rental/service/RentalService.java]: change "rated" to "rented"

**Review by DFraczek — CHANGES_REQUESTED** [review]
change function name

**Review by DFraczek — APPROVED** [review]
LGTM

**Comment by blackspyek** [comment]
image

**Pull request merged**